pkg/azurefile/controllerserver.go

```diff
@@ -1083,6 +1084,70 @@ func (d *Driver) ListSnapshots(ctx context.Context, req *csi.ListSnapshotsReques
 	return nil, status.Error(codes.Unimplemented, "")
 }
 
+// restoreSnapshot restores from a snapshot
+func (d *Driver) restoreSnapshot(ctx context.Context, dstAccountName string, req *csi.CreateVolumeRequest, accountKey string, shareOptions *fileclient.ShareOptions, storageEndpointSuffix string) error {
+	if shareOptions.Protocol == storage.EnabledProtocolsNFS {
+		return fmt.Errorf("protocol nfs is not supported for snapshot restore")
+	}
+	var sourceSnapshotID string
+	if req.GetVolumeContentSource() != nil && req.GetVolumeContentSource().GetSnapshot() != nil {
+		sourceSnapshotID = req.GetVolumeContentSource().GetSnapshot().GetSnapshotId()
+	}
+	resourceGroupName, srcAccountName, srcFileShareName, _, _, _, err := GetFileShareInfo(sourceSnapshotID) //nolint:dogsled
+	if err != nil {
+		return status.Error(codes.NotFound, err.Error())
+	}
+	dstFileShareName := shareOptions.Name
+	if srcFileShareName == "" || dstFileShareName == "" {
+		return fmt.Errorf("srcFileShareName(%s) or dstFileShareName(%s) is empty", srcFileShareName, dstFileShareName)
+	}
+
+	klog.V(2).Infof("generate sas token for src account(%s)", srcAccountName)
+	srcAccountSasToken, err := generateSASToken(srcAccountName, accountKey, storageEndpointSuffix, d.sasTokenExpirationMinutes)
+	if err != nil {
+		return azure.ErrorNotVmssInstance
+	}
+	klog.V(2).Infof("generate sas token for dst account(%s)", dstAccountName)
+	dstAccountSasToken, err := generateSASToken(dstAccountName, accountKey, storageEndpointSuffix, d.sasTokenExpirationMinutes)
+	if err != nil {
+		return err
+	}
+
+	timeAfter := time.After(waitForCopyTimeout)
+	timeTick := time.Tick(waitForCopyInterval)
+	srcPath := fmt.Sprintf("https://%s.file.%s/%s%s", srcAccountName, storageEndpointSuffix, srcFileShareName, srcAccountSasToken)
+	dstPath := fmt.Sprintf("https://%s.file.%s/%s%s", dstAccountName, storageEndpointSuffix, dstFileShareName, dstAccountSasToken)
+
+	jobState, percent, err := getAzcopyJob(dstFileShareName)
+	klog.V(2).Infof("azcopy job status: %s, copy percent: %s%%, error: %v", jobState, percent, err)
+	if jobState == AzcopyJobError || jobState == AzcopyJobCompleted {
+		return err
+	}
+	klog.V(2).Infof("begin to restore snapshot, copy fileshare %s to %s", srcFileShareName, dstFileShareName)
+	for {
+		select {
+		case <-timeTick:
+			jobState, percent, err := getAzcopyJob(dstFileShareName)
```

Comment:
can you combine the logic with copy volume, the `select ... case` logic is quite similar

Reply:
I think you could use same function for both restoreFromSansphot and copyVolume

Reply:
It's because srcPath is different from copy volume and snapshot has other logic for different storage account, should we combine the logic just for azcopy?

Reply:
combine the logic in a `copyFileShareByAzcopy` func